File: source python/Utilities/Position/Hex.py
import random
import pygame
from Application.Utilities.Position.Point import Point


class Hex:

    def __init__(self, x=0, y=0, z=None):
        if z is None:
            self.x = x
            self.z = y
            self.y = -self.x - self.z
        else:
            self.x, self.y, self.z = x, y, z

    def ToPoint(self):
        return Point(self.x, self.z)

    def __hash__(self):
        return int((self.x + self.z) * (self.x + self.z + 1) / 2 + self.z)

    def __str__(self):
        return "Hex (" + str(self.x) + ", " + str(self.y) + ", " + str(self.z) + ")"

    def __eq__(self, other):
        if not isinstance(other, Hex):
            return False
        return (self.x, self.y, self.z) == (other.x, other.y, other.z)

    def __ne__(self, other):
        return not (self == other)

    def __add__(self, other):
        return Hex(self.x + other.x, self.y + other.y, self.z + other.z)

    @staticmethod
    def GetRandom(width, height):
        return Hex(random.randint(0, width - 1), random.randint(0, height - 1))

    def Distance(self, hex):
        return (abs(self.x - hex.x) + abs(self.y - hex.y) + abs(self.z - hex.z)) / 2

    def IsInBoundary(self, width, height):
        return self.x >= 0 and self.x < width and self.z >= 0 and self.z < height

    def GetRandomNeighbour(self):
        return self + Hex.Neighbours()[random.randint(0, len(Hex.Neighbours()) - 1)]

    def GetAllNeighbours(self):
        absoluteNeighbours = Hex.Neighbours()
        result = list()
        for absoluteNeighbour in absoluteNeighbours:
            result.append(self + absoluteNeighbour)
        return result

    def GetNeighbouringCloserPosition(self, position):
        distance = self.Distance(position)
        if distance == 0:
            return self

        neighbours = self.GetAllNeighbours()
        for neighbour in neighbours:
            currentDistance = neighbour.Distance(position)
            if currentDistance < distance:
                return neighbour
        return self

    def GetMovement(self, key):
        if key == pygame.K_w:
            if self.z % 2 == 0:
                return self + Hex(-1, -1)
            return self + Hex(0, 1, -1)
        elif key == pygame.K_e:
            if self.z % 2 == 0:
                return self + Hex(0, -1)
            return self + Hex(1, 0, -1)
        elif key == pygame.K_d:
            return self + Hex(1, -1, 0)
        elif key == pygame.K_x:
            if self.z % 2 == 0:
                return self + Hex(0, 1)
            return self + Hex(1, 1)
        elif key == pygame.K_z:
            if self.z % 2 == 0:
                return self + Hex(-1, 1)
            return self + Hex(0, 1)
        elif key == pygame.K_a:
            return self + Hex(-1, 1, 0)
        else:
            return None


    @staticmethod
    def Neighbours():
        if not hasattr(Hex.Neighbours, "neighbours"):
            Hex.Neighbours.__neighbours = [Hex(0, -1), Hex(1, -1), Hex(1, 0), Hex(0, 1), Hex(-1, 1), Hex(-1, 0)]
        return Hex.Neighbours.__neighbours

    def Save(self, saver):
        saver.Insert(self.x)
        saver.Insert(self.y)
        saver.Insert(self.z)

    @staticmethod
    def Load(loader):
        return Hex(int(loader.Next()), int(loader.Next()), int(loader.Next()))

File: source java/utilities/position/Hex.java
package utilities.position;

import com.sun.glass.events.KeyEvent;
import utilities.savingsystem.Loader;
import utilities.savingsystem.Saver;

public class Hex extends Position
{
    private static Hex[] neighbours_ = {
        new Hex(0, -1), new Hex(1, -1), new Hex(1, 0),
        new Hex(0, 1), new Hex(-1, 1), new Hex(-1, 0)};
    
    public int x;
    public int y;
    public int z;
    
    public Hex()
    {
        x = y = z = 0;
    }
    
    public Hex(int x, int y, int z)
    {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    
    public Hex(int x, int y)
    {
        ToHex(x, y);
    }
    
    private void ToHex(int x, int y)
    {
        this.x = x;
        this.z = y;
        this.y = -this.x - this.z;
    }
    
    public Point ToPoint()
    {
        return new Point(x, z);
    }
    
    @Override
    public int Distance(Position other) 
    {
        if(!(other instanceof Hex))
            return -1;
        
        Hex otherHex = (Hex)other;
        
        return (Math.abs(x - otherHex.x) + Math.abs(y - otherHex.y) + Math.abs(z - otherHex.z)) / 2;
    }

    @Override
    public Hex Add(Position other) 
    {
        Hex hex = (Hex)(other);
        return new Hex(x + hex.x, y + hex.y, z + hex.z);
    }

    @Override
    public void Save(Saver saver) {
        saver.Print(Integer.toString(x));
        saver.Print(Integer.toString(y));
        saver.Print(Integer.toString(z));
    }

    @Override
    public Position Load(Loader loader) {
        x = loader.NextInt();
        y = loader.NextInt();
        z = loader.NextInt();
        return this;
    }

    @Override
    public Hex[] GetNeighbours() {
        Hex neighbours[] = new Hex[neighbours_.length];
        for(int i = 0; i < neighbours_.length; i++)
            neighbours[i] = this.Add(neighbours_[i]);
        return neighbours;
    }

    @Override
    public Position GetNeighbouringCloserPosition(Position destination) 
    {
        if(!(destination instanceof Hex))
                System.out.println("Gdzieś popełniłem bląd");
        
        int distance = this.Distance(destination);
        
        for(Hex neighbour : neighbours_)
        {
            int newDistance = (this.Add(neighbour)).Distance(destination);
            if(distance > newDistance)
                return this.Add(neighbour);
        }
        
        return this;
    }

    @Override
    public boolean IsInBoundaries(int width, int height) {
        return ToPoint().IsInBoundaries(width, height);
    }

    @Override
    public Hex SetRandom(int width, int height) {
        Point p =  (new Point()).SetRandom(width, height);
        ToHex(p.x, p.y);
        return this;
    }

    @Override
    public Hex GetMovement(int keyCode) {      
        switch(keyCode)
        {
            
            
            case KeyEvent.VK_W:
                if(this.ToPoint().y%2==0)
                    return this.Add(new Hex(-1, -1));
                return this.Add(new Hex(0, 1, -1));
            case KeyEvent.VK_E:
                if(this.ToPoint().y%2==0)
                    return this.Add(new Hex(0, -1));
                return this.Add(new Hex(1, 0, -1));
            case KeyEvent.VK_D:
                return this.Add(new Hex(1, -1, 0));
            case KeyEvent.VK_X:
                if(this.ToPoint().y%2==0)
                    return this.Add(new Hex(0, 1));
                return this.Add(new Hex(1, 1));
            case KeyEvent.VK_Z:
                if(this.ToPoint().y%2==0)
                    return this.Add(new Hex(-1, 1));
                return this.Add(new Hex(0, 1));
            case KeyEvent.VK_A:
                return this.Add(new Hex(-1, 1, 0));
        }
        
        return null;
    }

    @Override
    public Hex GetRandomNeighbour() {
        return this.Add(neighbours_[generator.nextInt(neighbours_.length)]);
    }

    @Override
    public int hashCode() 
    {
        return this.ToPoint().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        
        final Hex other = (Hex) obj;
        return (other.x == this.x && other.y == this.y && other.z == this.z);
    }

    @Override
    public String toString() {
        return ToPoint().toString();
        //return "Hex (" + x + ", " + y + ", " + z + ")";
    }

}
